File: core/Sources/ChotkiCore/Scheduling/Scheduler.swift
import Foundation

/// Decides what should be reminded, and when.
///
/// Pure: given rules, activations, occurrences and a day, it returns the
/// reminders for that day. It does not fire anything, does not sleep, and does
/// not know what platform it is on — the `Notifier` only shows what this
/// decides. That is what lets a simulated month run headlessly in CI.
public struct Scheduler: Sendable {

    private let engine: RecurrenceEngine
    private let policy: ReminderPolicy
    private let timeZone: TimeZone

    public init(
        engine: RecurrenceEngine = RecurrenceEngine(),
        policy: ReminderPolicy = .default,
        timeZone: TimeZone = .current
    ) {
        self.engine = engine
        self.policy = policy
        self.timeZone = timeZone
    }

    /// Statuses that end a day's reminders. Completing is the obvious one;
    /// skipping and cancelling must silence it too, or standing something down
    /// would keep buzzing about it.
    private static let settled: Set<OccurrenceStatus> = [
        .completed, .completedLate, .skipped, .cancelled, .moved
    ]

    /// Every reminder for one day, in fire order.
    public func plan(
        rules: [Rule],
        activations: [Activation],
        occurrences: [Occurrence],
        on date: CalendarDate
    ) -> [PlannedNotification] {
        let settledRuleIDs = Set(
            occurrences
                .filter { $0.date == date && Scheduler.settled.contains($0.status) }
                .map(\.ruleID)
        )

        var planned: [PlannedNotification] = []
        for rule in rules where !rule.isArchived && !settledRuleIDs.contains(rule.id) {
            let due = engine.dueDates(
                rule: rule, activations: activations, from: date, through: date
            )
            guard !due.isEmpty else { continue }
            planned += reminders(for: rule, on: date)
        }
        return planned.sorted { $0.fireAt < $1.fireAt }
    }

    /// Reminders still ahead of `now` — what a driver would actually arm.
    public func pending(
        rules: [Rule],
        activations: [Activation],
        occurrences: [Occurrence],
        on date: CalendarDate,
        after now: Date
    ) -> [PlannedNotification] {
        plan(rules: rules, activations: activations, occurrences: occurrences, on: date)
            .filter { $0.fireAt > now }
    }

    /// The ids to cancel when a rule is completed, paused or archived. Callers
    /// need not know how many reminders were armed.
    public func cancellationIDs(
        ruleID: UUID, date: CalendarDate, rules: [Rule], activations: [Activation]
    ) -> [String] {
        guard let rule = rules.first(where: { $0.id == ruleID }) else { return [] }
        return reminders(for: rule, on: date).map(\.id)
    }

    // MARK: building reminders

    private func reminders(for rule: Rule, on date: CalendarDate) -> [PlannedNotification] {
        if let time = rule.timeOfDay {
            return timedReminder(rule: rule, date: date, time: time).map { [$0] } ?? []
        }
        return untimedReminders(rule: rule, date: date)
    }

    private func timedReminder(
        rule: Rule, date: CalendarDate, time: TimeOfDay
    ) -> PlannedNotification? {
        // A wall-clock time the day does not have — 02:30 on a spring-forward
        // morning — yields no instant, so no reminder rather than a wrong one.
        guard let dueAt = date.dueInstant(at: time, in: timeZone) else { return nil }
        let fireAt = dueAt.addingTimeInterval(-policy.leadTime)

        // Quiet hours deliberately do NOT apply here. They exist to stop
        // unsolicited hourly nagging, not to suppress a reminder the user set
        // themselves: morning prayers at 06:30 have a lead time of 06:20, which
        // falls inside the default quiet window, and silencing it would make
        // the app useless for exactly the rule people most want kept.

        return PlannedNotification(
            id: "\(PlannedNotification.occurrenceKey(ruleID: rule.id, date: date)):due",
            ruleID: rule.id,
            date: date,
            fireAt: fireAt,
            request: NotificationRequest(
                id: "\(PlannedNotification.occurrenceKey(ruleID: rule.id, date: date)):due",
                title: rule.title,
                // Neutral by construction: what is due and when, never how long
                // it has been outstanding. See the Tone section in design.md.
                body: "At \(format(time))",
                actions: [.markComplete, .snooze]
            )
        )
    }

    private func untimedReminders(rule: Rule, date: CalendarDate) -> [PlannedNotification] {
        let cap = policy.untimedCap > 0 ? policy.untimedCap : Int.max

        // Every slot the quiet window leaves open, in order.
        let stepMinutes = max(Int(policy.untimedInterval / 60), 1)
        var waking: [TimeOfDay] = []
        var minute = 0
        while minute < 24 * 60 {
            defer { minute += stepMinutes }
            guard let time = TimeOfDay(hour: minute / 60, minute: minute % 60) else { continue }
            if !policy.quietHours.contains(time) { waking.append(time) }
        }
        guard !waking.isEmpty else { return [] }

        let chosen: [TimeOfDay]
        switch policy.spacing {
        case .hourly:
            chosen = Array(waking.prefix(cap))
        case .spreadAcrossDay:
            let wanted = min(cap, waking.count)
            if wanted <= 1 {
                chosen = Array(waking.prefix(wanted))
            } else {
                // Evenly spaced across the waking window, first and last included.
                let stride = Double(waking.count - 1) / Double(wanted - 1)
                chosen = (0..<wanted).map { waking[Int((Double($0) * stride).rounded())] }
            }
        }

        var result: [PlannedNotification] = []
        for time in chosen {
            guard let fireAt = date.dueInstant(at: time, in: timeZone) else { continue }
            let index = result.count
            let id = "\(PlannedNotification.occurrenceKey(ruleID: rule.id, date: date)):\(index)"
            result.append(
                PlannedNotification(
                    id: id, ruleID: rule.id, date: date, fireAt: fireAt,
                    request: NotificationRequest(
                        id: id,
                        title: rule.title,
                        body: "Today",
                        actions: [.markComplete, .snooze]
                    )
                )
            )
        }
        return result
    }

    private func format(_ time: TimeOfDay) -> String {
        String(format: "%02d:%02d", time.hour, time.minute)
    }
}

File: android/core/src/main/kotlin/org/chotki/core/scheduling/Scheduler.kt
package org.chotki.core.scheduling

import org.chotki.core.Activation
import org.chotki.core.CalendarDate
import org.chotki.core.NotificationAction
import org.chotki.core.NotificationRequest
import org.chotki.core.Occurrence
import org.chotki.core.OccurrenceStatus
import org.chotki.core.RecurrenceEngine
import org.chotki.core.Rule
import org.chotki.core.TimeOfDay
import java.time.Instant
import java.time.ZoneId
import java.util.UUID
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * Decides what should be reminded, and when.
 *
 * Pure: given rules, activations, occurrences and a day, it returns the
 * reminders for that day. It does not fire anything, does not sleep, and does
 * not know what platform it is on — the notifier only shows what this decides.
 * That is what lets a simulated month run headlessly in CI.
 */
class Scheduler(
    private val engine: RecurrenceEngine = RecurrenceEngine(),
    private val policy: ReminderPolicy = ReminderPolicy.DEFAULT,
    private val zone: ZoneId = ZoneId.systemDefault(),
) {
    private companion object {
        /**
         * Statuses that end a day's reminders. Completing is the obvious one;
         * skipping and cancelling must silence it too, or standing something
         * down would keep buzzing about it.
         */
        val SETTLED = setOf(
            OccurrenceStatus.COMPLETED,
            OccurrenceStatus.COMPLETED_LATE,
            OccurrenceStatus.SKIPPED,
            OccurrenceStatus.CANCELLED,
            OccurrenceStatus.MOVED,
        )
    }

    /** Every reminder for one day, in fire order. */
    fun plan(
        rules: List<Rule>,
        activations: List<Activation>,
        occurrences: List<Occurrence>,
        on: CalendarDate,
    ): List<PlannedNotification> {
        // The master switch silences everything and changes nothing else: rules
        // stay due, and scoring never sees this.
        if (!policy.notificationsEnabled) return emptyList()

        val settledRuleIDs = occurrences
            .filter { it.date == on && it.status in SETTLED }
            .map { it.ruleID }
            .toSet()

        return rules
            .filterNot { it.isArchived || it.id in settledRuleIDs }
            .filter { it.effectiveReminders.enabled }
            .filter { engine.dueDates(it, activations, on, on).isNotEmpty() }
            .flatMap { reminders(it, on) }
            .sortedBy { it.fireAt }
    }

    /** Reminders still ahead of [after] — what a driver would actually arm. */
    fun pending(
        rules: List<Rule>,
        activations: List<Activation>,
        occurrences: List<Occurrence>,
        on: CalendarDate,
        after: Instant,
    ): List<PlannedNotification> =
        plan(rules, activations, occurrences, on).filter { it.fireAt > after }

    /**
     * The ids to cancel when a rule is completed, paused or archived. Callers
     * need not know how many reminders were armed.
     */
    fun cancellationIDs(
        ruleID: UUID,
        date: CalendarDate,
        rules: List<Rule>,
    ): List<String> {
        val rule = rules.firstOrNull { it.id == ruleID } ?: return emptyList()
        return reminders(rule, date).map { it.id }
    }

    // MARK: building reminders

    private fun reminders(rule: Rule, date: CalendarDate): List<PlannedNotification> {
        val time = rule.timeOfDay
        return if (time != null) timedReminders(rule, date, time) else untimedReminders(rule, date)
    }

    /**
     * One reminder per configured lead. More than one is allowed — an hour
     * before to get ready, ten minutes before to actually leave.
     */
    private fun timedReminders(
        rule: Rule,
        date: CalendarDate,
        time: TimeOfDay,
    ): List<PlannedNotification> {
        // A wall-clock time the day does not have — 02:30 on a spring-forward
        // morning — yields no instant, so no reminder rather than a wrong one.
        val dueAt = date.dueInstant(time, zone) ?: return emptyList()

        val configured = rule.effectiveReminders.leads
        val leads = configured.ifEmpty { listOf(policy.defaultLead) }
        val key = PlannedNotification.occurrenceKey(rule.id, date)

        return leads.sortedWith(ReminderLead.BY_LEAD).mapNotNull { lead ->
            val fireAt = if (lead == ReminderLead.THE_EVENING_BEFORE) {
                // A fixed, predictable hour the evening before, so something you
                // travel to can be prepared for the night before.
                date.plusDays(-1).dueInstant(TimeOfDay.of(20, 0)!!, zone) ?: return@mapNotNull null
            } else {
                dueAt.minusSeconds(lead.seconds)
            }

            // Quiet hours deliberately do NOT apply to a rule the user gave a
            // time to. They exist to stop unsolicited repetition, not to silence
            // a reminder that was asked for: with the default window ending
            // 06:30, a 06:30 rule warns at 06:20 and must still arrive.
            val id = "$key:lead${lead.minutes}"
            PlannedNotification(
                id = id,
                ruleID = rule.id,
                date = date,
                fireAt = fireAt,
                request = NotificationRequest(
                    id = id,
                    title = rule.title,
                    // Neutral by construction: what is due and when, never how
                    // long it has been outstanding.
                    body = body(lead, time),
                    actions = listOf(NotificationAction.MARK_COMPLETE, NotificationAction.SNOOZE),
                ),
            )
        }
    }

    private fun body(lead: ReminderLead, time: TimeOfDay): String {
        val at = "%02d:%02d".format(time.hour, time.minute)
        return if (lead == ReminderLead.THE_EVENING_BEFORE) "Tomorrow at $at" else "At $at"
    }

    private fun untimedReminders(rule: Rule, date: CalendarDate): List<PlannedNotification> {
        val cap = if (policy.untimedCap > 0) policy.untimedCap else Int.MAX_VALUE

        // Every slot the quiet window leaves open, in order.
        val stepMinutes = maxOf((policy.untimedIntervalSeconds / 60).toInt(), 1)
        val waking = mutableListOf<TimeOfDay>()
        var minute = 0
        while (minute < 24 * 60) {
            val time = TimeOfDay.of(minute / 60, minute % 60)
            if (time != null && !policy.quietHours.contains(time)) waking.add(time)
            minute += stepMinutes
        }
        if (waking.isEmpty()) return emptyList()

        val chosen: List<TimeOfDay> = when (policy.spacing) {
            UntimedSpacing.HOURLY -> waking.take(cap)
            UntimedSpacing.SPREAD_ACROSS_DAY -> {
                val wanted = min(cap, waking.size)
                if (wanted <= 1) {
                    waking.take(wanted)
                } else {
                    // Evenly spaced across the waking window, first and last
                    // included.
                    val stride = (waking.size - 1).toDouble() / (wanted - 1).toDouble()
                    (0 until wanted).map { waking[(it * stride).roundToInt()] }
                }
            }
        }

        val result = mutableListOf<PlannedNotification>()
        for (time in chosen) {
            val fireAt = date.dueInstant(time, zone) ?: continue
            val id = "${PlannedNotification.occurrenceKey(rule.id, date)}:${result.size}"
            result.add(
                PlannedNotification(
                    id = id,
                    ruleID = rule.id,
                    date = date,
                    fireAt = fireAt,
                    request = NotificationRequest(
                        id = id,
                        title = rule.title,
                        body = "Today",
                        actions = listOf(
                            NotificationAction.MARK_COMPLETE, NotificationAction.SNOOZE,
                        ),
                    ),
                ),
            )
        }
        return result
    }
}
